axum/src/docs/routing/with_state.md

```diff
@@ -105,6 +105,40 @@ receives. That means it is not suitable for holding state derived from a
 request, such as authorization data extracted in a middleware. Use [`Extension`]
 instead for such data.
 
+# Mutable state
+
+As state is used across routes it is not mutable by default. To mutate a state
+you will need to use an `Arc<Mutex>` or similar. Note that the Mutex must
+support async locks.
```

Comment:
You don't actually need an asynchronous mutex. Synchronous is fine and even recommended. An async mutex is only needed if the lock is held across `.await` points.

See also https://docs.rs/tokio/1.25.0/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use

Reply:
Yes. There are many caveats when it comes to shared mutable state in async. Therefore I'm hesitant to include examples of it in our docs since users might see it as the "axum approved way" without understanding what they're getting themselves into.

Reply:
Thanks for the feedback and the link. I mistakenly made the assumption about blocking mutex being required as my own project did require it.